File: list.go
package main

import (
	"time"
	"log"
	"strconv"
	"github.com/hsjunnesson/top-five/Godeps/_workspace/src/github.com/gin-gonic/gin"
	"github.com/hsjunnesson/top-five/Godeps/_workspace/src/gopkg.in/redis.v3"
)

type List struct {
	Id          string
	Title       string
	CreatedAt   time.Time
	ListItems   [5]string
}

type Lists []List

// Creates a List in the connected redis.
// Returns the id of the List, or error.
func CreateListDB(client *redis.Client, title string, listItems [5]string) (string, error) {
	timestamp := time.Now().Format(time.RFC3339)
	
	id, err := client.Incr("next_list_id").Result()
	if err != nil {
		log.Printf("Couldn't increment next_list_id", err)
		return "", err
	}
	
	strId := strconv.FormatInt(id, 10)
	
	if client.HMSet(
		"list:" + strId,
		"id", strId,
		"title", title,
		"created_at", timestamp,		
	).Err() != nil {
		log.Printf("Couldn't HMSet list:" + strId, err)
		return "", err
	}
	
	for _, val := range listItems {
		if client.RPush("list_items:" + strId, val).Err() != nil {
			log.Printf("Couldn't RPush list_items:" + strId, err)
			return "", err
		}
	}
	
	if client.LPush("lists", strId).Err() != nil {
		log.Printf("Couldn't LPush lists " + strId, err)
		return "", err
	}
	
	return strId, nil
}

// Fetches a List from the connected redis.
func GetListDB(client *redis.Client, id string) (*List, error) {
	parts, err := client.HMGet("list:" + id, "id", "title", "created_at").Result()
	if err != nil {
		log.Printf("Could not HMGet list:" + id, err)
		return nil, err
	}

	id, err := parts[0].(string)
	if err != nil {
		return nil, err
	}

	title, err := parts[1].(string)
	if err != nil {
		return nil, err
	}

		
	
	var items = [5]string{"one", "two", "three", "four", "five"}
	list := List{
		Id: parts[0],
		Title: parts[1],
		CreatedAt: time.New(parts[2]),
		ListItems: items,
	}

	return list, nil
}

func ListsList(c *gin.Context) {
	c.JSON(200, "Hej")
}

func ListsDetail(c *gin.Context) {
//	list_id := c.Params.ByName("id")
//	_, _ := strconv.Atoi(list_id)
//	list := lists()[0]
	//content := gin.H{"title": list.Title, "content": list.Items}
	c.JSON(200, "Hoj")
}

File: src/top_five_things/db/list.clj
(ns top-five-things.db.list
  (:import [java.util Date])
  (:require [taoensso.carmine :as car]
            [clj-time.core :as time]
            [clj-time.format :as timef])
  (:use [top-five-things.db.core :only [wcar*]]))

(defn create-list!
  "Creates a list in the connected redis.
   Returns the id of the list or throws an error."
  [title items]
  (let [id (str (wcar* (car/incr "next_list_id")))
        created-at (timef/unparse (timef/formatters :date-time) (time/now))]
    (wcar*
     (car/hmset (str "list:" id)
                 :id id
                 :title title
                 :created-at created-at)
     (doseq [item items] (car/rpush (str "list_items:" id) item))
     (car/lpush "lists" id))
    id))

(defn get-list [id]
  "Returns a list from the connected redis."
  (let [hash-key (str "list:" id)]
    {:id id
     :title (wcar* (car/hget hash-key :title))
     :created-at (timef/parse (timef/formatters :date-time) (wcar* (car/hget hash-key :created-at)))
     :items (wcar* (car/lrange (str "list_items:" id) 0 -1))}))

